File: internal/forms/field.go
package forms

import (
	"fmt"
	"strings"
)

// FieldType represents how a field's value is determined.
type FieldType int

const (
	UserInput  FieldType = iota // needs taxpayer input
	Computed                    // calculated from other fields
	Lookup                      // from tax tables
	PriorYear                   // carried from last year
	FederalRef                  // state form referencing federal field
)

// Jurisdiction represents the taxing authority for a form.
type Jurisdiction int

const (
	Federal Jurisdiction = iota
	StateCA
)

// DepValues is passed to Compute functions for accessing dependency values.
type DepValues struct {
	values    map[string]float64
	strValues map[string]string
	taxYear   int
}

// NewDepValues creates a DepValues from the given maps and tax year.
func NewDepValues(values map[string]float64, strValues map[string]string, taxYear int) DepValues {
	if values == nil {
		values = make(map[string]float64)
	}
	if strValues == nil {
		strValues = make(map[string]string)
	}
	return DepValues{
		values:    values,
		strValues: strValues,
		taxYear:   taxYear,
	}
}

// Get returns the float64 value for the given key, or 0 if not found.
func (d DepValues) Get(key string) float64 {
	return d.values[key]
}

// GetStrict returns the float64 value for the given key, or an error if the key doesn't exist.
func (d DepValues) GetStrict(key string) (float64, error) {
	v, ok := d.values[key]
	if !ok {
		return 0, fmt.Errorf("dependency key %q not found in DepValues", key)
	}
	return v, nil
}

// Keys returns all available numeric keys in the DepValues.
func (d DepValues) Keys() []string {
	keys := make([]string, 0, len(d.values))
	for k := range d.values {
		keys = append(keys, k)
	}
	return keys
}

// GetString returns the string value for the given key, or "" if not found.
func (d DepValues) GetString(key string) string {
	return d.strValues[key]
}

// SumAll sums all float64 values whose keys match the given wildcard pattern.
// The pattern supports '*' as a wildcard that matches any substring within a
// single segment. For example, "w2:*:wages" matches "w2:employer1:wages" and
// "w2:employer2:wages".
func (d DepValues) SumAll(pattern string) float64 {
	var sum float64
	for k, v := range d.values {
		if matchWildcard(pattern, k) {
			sum += v
		}
	}
	return sum
}

// SumAllWhere sums all float64 values whose keys match valuePattern, but only
// for instances where the corresponding filterPattern key has the given
// filterValue as its string value. This enables filtering wildcard sums by
// a categorical field (e.g., summing proceeds only for short-term transactions).
//
// Example: SumAllWhere("1099b:*:proceeds", "1099b:*:term", "short")
// sums proceeds for all 1099b instances where term == "short".
func (d DepValues) SumAllWhere(valuePattern, filterPattern, filterValue string) float64 {
	var sum float64
	for k, v := range d.values {
		if !matchWildcard(valuePattern, k) {
			continue
		}
		// Extract the instance segment from the matched key and build
		// the corresponding filter key.
		filterKey := buildCorrespondingKey(valuePattern, filterPattern, k)
		if filterKey == "" {
			continue
		}
		if sv, ok := d.strValues[filterKey]; ok && sv == filterValue {
			sum += v
		}
	}
	return sum
}

// buildCorrespondingKey takes a valuePattern, a filterPattern, and a concrete
// key that matched valuePattern, and returns the concrete key that corresponds
// to filterPattern with the same wildcard segment.
//
// Example: buildCorrespondingKey("1099b:*:proceeds", "1099b:*:term", "1099b:1:proceeds")
// returns "1099b:1:term"
func buildCorrespondingKey(valuePattern, filterPattern, concreteKey string) string {
	// Find the wildcard segment by splitting both pattern and concrete key
	vParts := strings.Split(valuePattern, "*")
	if len(vParts) != 2 {
		return "" // only single-wildcard patterns supported
	}
	prefix := vParts[0]
	suffix := vParts[1]

	if !strings.HasPrefix(concreteKey, prefix) || !strings.HasSuffix(concreteKey, suffix) {
		return ""
	}

	// Extract the wildcard segment
	wildcardSeg := concreteKey[len(prefix) : len(concreteKey)-len(suffix)]

	// Build the filter key by replacing * with the wildcard segment
	return strings.Replace(filterPattern, "*", wildcardSeg, 1)
}

// TaxYear returns the tax year for this computation context.
func (d DepValues) TaxYear() int {
	return d.taxYear
}

// matchWildcard checks if s matches a pattern containing '*' wildcards.
// Each '*' matches zero or more characters.
func matchWildcard(pattern, s string) bool {
	// Split pattern on '*' and check segments appear in order
	parts := strings.Split(pattern, "*")
	if len(parts) == 1 {
		return pattern == s
	}

	idx := 0
	for i, part := range parts {
		if part == "" {
			continue
		}
		pos := strings.Index(s[idx:], part)
		if pos < 0 {
			return false
		}
		if i == 0 && pos != 0 {
			// First part must match at start if it's non-empty
			return false
		}
		idx += pos + len(part)
	}
	// If the last part is non-empty, the string must end with it
	if last := parts[len(parts)-1]; last != "" {
		return strings.HasSuffix(s, last)
	}
	return true
}

// FieldValueType indicates the data type of a field's value.
type FieldValueType int

const (
	NumericValue FieldValueType = iota // default: numeric/currency
	StringValue                        // text input (names, SSN, EIN, etc.)
	IntegerValue                       // whole number (counts, not currency)
)

// FieldDef defines a single field on a tax form.
type FieldDef struct {
	Line       string
	Type       FieldType
	ValueType  FieldValueType // NumericValue or StringValue
	Label      string
	Prompt     string                  // human-readable question (for UserInput)
	DependsOn  []string                // field keys this depends on (form_id:line format)
	Options    []string                // for enum-type UserInput fields
	Compute    func(DepValues) float64 // for Computed/FederalRef/Lookup fields
	ComputeStr func(DepValues) string  // for string-valued computed fields
}

// FormDef defines a complete tax form.
type FormDef struct {
	ID            FormID
	Name          string
	Jurisdiction  Jurisdiction
	TaxYears      []int
	Fields        []FieldDef
	QuestionGroup string // e.g., "personal", "income_w2", "expat", "ca"
	QuestionOrder int    // sort order within group (lower = earlier)

	fieldIndex map[string]int // lazily built: line -> index in Fields
}

// FieldByLine returns the FieldDef for the given line, or nil if not found.
// The first call builds an internal index for O(1) lookups on subsequent calls.
func (f *FormDef) FieldByLine(line string) *FieldDef {
	if f.fieldIndex == nil {
		f.fieldIndex = make(map[string]int, len(f.Fields))
		for i := range f.Fields {
			f.fieldIndex[f.Fields[i].Line] = i
		}
	}
	idx, ok := f.fieldIndex[line]
	if !ok {
		return nil
	}
	return &f.Fields[idx]
}

// FieldKey returns the fully qualified key for a field: "form_id:line".
func FieldKey(formID FormID, line string) string {
	return fmt.Sprintf("%s:%s", string(formID), line)
}

// FK is a convenience for building field keys with a typed FormID and line.
// It is identical to FieldKey but shorter for use in DependsOn lists and map lookups.
func FK(formID FormID, line string) string {
	return string(formID) + ":" + line
}

File: api/src/domain/field.rs
use std::collections::HashMap;
use std::fmt;

pub use crate::domain::wildcard::{build_corresponding_key, match_wildcard};

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum FieldType {
    UserInput,
    Computed,
    Lookup,
    PriorYear,
    FederalRef,
}

impl fmt::Display for FieldType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            FieldType::UserInput => write!(f, "UserInput"),
            FieldType::Computed => write!(f, "Computed"),
            FieldType::Lookup => write!(f, "Lookup"),
            FieldType::PriorYear => write!(f, "PriorYear"),
            FieldType::FederalRef => write!(f, "FederalRef"),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum FieldValueType {
    Numeric,
    String,
    Integer,
}

impl Default for FieldValueType {
    fn default() -> Self {
        FieldValueType::Numeric
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Jurisdiction {
    Federal,
    StateCA,
}

impl fmt::Display for Jurisdiction {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Jurisdiction::Federal => write!(f, "Federal"),
            Jurisdiction::StateCA => write!(f, "CA"),
        }
    }
}

// ---------------------------------------------------------------------------
// DepValues -- dependency values passed to compute closures
// ---------------------------------------------------------------------------

#[derive(Debug, Clone)]
pub struct DepValues {
    pub num: HashMap<String, f64>,
    pub str_vals: HashMap<String, String>,
    pub tax_year: i32,
}

impl DepValues {
    pub fn new(
        num: HashMap<String, f64>,
        str_vals: HashMap<String, String>,
        tax_year: i32,
    ) -> Self {
        Self {
            num,
            str_vals,
            tax_year,
        }
    }

    /// Get a numeric value, returning 0.0 if not found.
    pub fn get(&self, key: &str) -> f64 {
        self.num.get(key).copied().unwrap_or(0.0)
    }

    /// Get a numeric value, returning an error if not found.
    pub fn get_strict(&self, key: &str) -> Result<f64, String> {
        self.num
            .get(key)
            .copied()
            .ok_or_else(|| format!("dependency key {:?} not found in DepValues", key))
    }

    /// Get a string value, returning empty string if not found.
    pub fn get_string(&self, key: &str) -> String {
        self.str_vals.get(key).cloned().unwrap_or_default()
    }

    /// Return all keys in the numeric map.
    pub fn keys(&self) -> Vec<&String> {
        self.num.keys().collect()
    }

    /// Sum all numeric values whose keys match the given wildcard pattern.
    pub fn sum_all(&self, pattern: &str) -> f64 {
        self.num
            .iter()
            .filter(|(k, _)| match_wildcard(pattern, k))
            .map(|(_, v)| *v)
            .sum()
    }

    /// Sum all numeric values whose keys match value_pattern, but only for
    /// instances where the corresponding filter_pattern key has filter_value
    /// as its string value.
    ///
    /// Example: sum_all_where("1099b:*:proceeds", "1099b:*:term", "short")
    /// sums proceeds for all 1099b instances where term == "short".
    pub fn sum_all_where(
        &self,
        value_pattern: &str,
        filter_pattern: &str,
        filter_value: &str,
    ) -> f64 {
        let mut sum = 0.0;
        for (k, v) in &self.num {
            if !match_wildcard(value_pattern, k) {
                continue;
            }
            if let Some(filter_key) =
                build_corresponding_key(value_pattern, filter_pattern, k)
            {
                if let Some(sv) = self.str_vals.get(&filter_key) {
                    if sv == filter_value {
                        sum += v;
                    }
                }
            }
        }
        sum
    }

    /// Return the tax year.
    pub fn tax_year(&self) -> i32 {
        self.tax_year
    }
}

// ---------------------------------------------------------------------------
// FieldDef
// ---------------------------------------------------------------------------

pub struct FieldDef {
    pub line: String,
    pub field_type: FieldType,
    pub value_type: FieldValueType,
    pub label: String,
    pub prompt: String,
    pub depends_on: Vec<String>,
    pub options: Vec<String>,
    pub compute: Option<Box<dyn Fn(&DepValues) -> f64 + Send + Sync>>,
    pub compute_str: Option<Box<dyn Fn(&DepValues) -> String + Send + Sync>>,
}

impl fmt::Debug for FieldDef {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("FieldDef")
            .field("line", &self.line)
            .field("field_type", &self.field_type)
            .field("value_type", &self.value_type)
            .field("label", &self.label)
            .field("depends_on", &self.depends_on)
            .finish()
    }
}

impl FieldDef {
    pub fn new_user_input(line: &str, label: &str, prompt: &str) -> Self {
        Self {
            line: line.to_string(),
            field_type: FieldType::UserInput,
            value_type: FieldValueType::Numeric,
            label: label.to_string(),
            prompt: prompt.to_string(),
            depends_on: Vec::new(),
            options: Vec::new(),
            compute: None,
            compute_str: None,
        }
    }

    pub fn new_computed(
        line: &str,
        label: &str,
        deps: Vec<String>,
        compute: Box<dyn Fn(&DepValues) -> f64 + Send + Sync>,
    ) -> Self {
        Self {
            line: line.to_string(),
            field_type: FieldType::Computed,
            value_type: FieldValueType::Numeric,
            label: label.to_string(),
            prompt: String::new(),
            depends_on: deps,
            options: Vec::new(),
            compute: Some(compute),
            compute_str: None,
        }
    }
}

// ---------------------------------------------------------------------------
// FormDef
// ---------------------------------------------------------------------------

pub struct FormDef {
    pub id: String,
    pub name: String,
    pub jurisdiction: Jurisdiction,
    pub tax_years: Vec<i32>,
    pub fields: Vec<FieldDef>,
    pub question_group: String,
    pub question_order: i32,
}

impl fmt::Debug for FormDef {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("FormDef")
            .field("id", &self.id)
            .field("name", &self.name)
            .field("jurisdiction", &self.jurisdiction)
            .field("fields_count", &self.fields.len())
            .finish()
    }
}

impl FormDef {
    /// Returns the FieldDef for the given line, or None if not found.
    pub fn field_by_line(&self, line: &str) -> Option<&FieldDef> {
        self.fields.iter().find(|f| f.line == line)
    }
}

// ---------------------------------------------------------------------------
// Key helpers
// ---------------------------------------------------------------------------

/// Build a field key from form ID and line, e.g. field_key("1040", "line1") => "1040:line1"
pub fn field_key(form_id: &str, line: &str) -> String {
    format!("{}:{}", form_id, line)
}

/// Shorthand alias for field_key.
pub fn fk(form_id: &str, line: &str) -> String {
    field_key(form_id, line)
}
